Fixed #35628 -- Allowed compatible GeneratedFields for ModelAdmin.date_hierarchy.

john-parton opened the pull request:
Fixed #35628, allow GeneratedField with appropriate output_field for date_hierarchy.

# Trac ticket number

This is the basic fix for https://code.djangoproject.com/ticket/35628# , but without tests/docs. Will add tests/docs if approach is deemed appropriate.

# Branch description
Essentially, there's a check to make sure that `date_hierarchy` is a `DateField` or `DateTimeField`, but there is no technical reason to prohibit a `GeneratedField` with `output_field` set to `DateField` or `DateTimeField`.

This adjusts the check to allow that specific case through.

# Checklist
- [x] This PR targets the `main` branch. <!-- Backports will be evaluated and done by mergers, when necessary. -->
- [x] The commit message is written in past tense, mentions the ticket number, and ends with a period.
- [x] I have checked the "Has patch" ticket flag in the Trac system.
- [x] I have added or updated relevant tests.
- [x] I have added or updated relevant docs, including release notes if applicable.
- [ ] I have attached screenshots in both light and dark modes for any UI changes.

john-parton commented:
As an alternate implementation, we could do something like

```python
if field.get_internal_type() in {"DateField", "DateTimeField"}:
    ...
 ```
 
 This would also permit user defined date fields that don't explicitly subclass `DateField` or `DateTimeField` so long as the user defines the `get_internal_type` method.

charettes commented on django/contrib/admin/checks.py:
```diff
@@ -1184,16 +1184,21 @@ def _check_date_hierarchy(self, obj):
-                else:
+                supported_fields = (models.DateField, models.DateTimeField)
+
+                if isinstance(field, supported_fields) or (
+                    isinstance(field, models.GeneratedField)
+                    and isinstance(field.output_field, supported_fields)
+                ):
```
I think that checking for `output_field` like you did over `get_internal_type` is the better solution.

john-parton replied:
Sounds good. I'll see if I can add tests/docs later today. Docs should be trivial. I'll reach out if I get stuck on tests.

sarahboyce replied:
Sorry to follow up on this, as this is a release blocker we will need to have this merged by August 6th (Tuesday)
<details><summary>Here are some tests in case this helps</summary>

```diff
diff --git a/tests/modeladmin/test_checks.py b/tests/modeladmin/test_checks.py
index f767a6c92b..1206da2892 100644
--- a/tests/modeladmin/test_checks.py
+++ b/tests/modeladmin/test_checks.py
@@ -4,16 +4,17 @@ from django.contrib.admin import BooleanFieldListFilter, SimpleListFilter
 from django.contrib.admin.options import VERTICAL, ModelAdmin, TabularInline
 from django.contrib.admin.sites import AdminSite
 from django.core.checks import Error
+from django.db import models
 from django.db.models import CASCADE, F, Field, ForeignKey, ManyToManyField, Model
 from django.db.models.functions import Upper
 from django.forms.models import BaseModelFormSet
-from django.test import SimpleTestCase
+from django.test import TestCase, skipUnlessDBFeature
 from django.test.utils import isolate_apps
 
 from .models import Band, Song, User, ValidationTestInlineModel, ValidationTestModel
 
 
-class CheckTestCase(SimpleTestCase):
+class CheckTestCase(TestCase):
     def assertIsInvalid(
         self,
         model_admin,
@@ -1035,6 +1036,22 @@ class DateHierarchyCheckTests(CheckTestCase):
 
         self.assertIsValid(TestModelAdmin, ValidationTestModel)
 
+    @skipUnlessDBFeature("supports_stored_generated_columns")
+    @isolate_apps("modeladmin")
+    def test_valid_case_generated_field(self):
+        class TestModel(Model):
+            date = models.DateTimeField()
+            date_copy = models.GeneratedField(
+                expression=F("date"),
+                output_field=models.DateTimeField(),
+                db_persist=True,
+            )
+
+        class TestModelAdmin(ModelAdmin):
+            date_hierarchy = "date_copy"
+
+        self.assertIsValid(TestModelAdmin, TestModel)
+
     def test_related_valid_case(self):
         class TestModelAdmin(ModelAdmin):
             date_hierarchy = "band__sign_date"
@@ -1052,6 +1069,26 @@ class DateHierarchyCheckTests(CheckTestCase):
             "admin.E128",
         )
 
+    @skipUnlessDBFeature("supports_stored_generated_columns")
+    @isolate_apps("modeladmin")
+    def test_related_invalid_field_type_generated_field(self):
+        class TestModel(Model):
+            generated = models.GeneratedField(
+                expression=models.Value(1),
+                output_field=models.IntegerField(),
+                db_persist=True,
+            )
+
+        class TestModelAdmin(ModelAdmin):
+            date_hierarchy = "generated"
+
+        self.assertIsInvalid(
+            TestModelAdmin,
+            TestModel,
+            "The value of 'date_hierarchy' must be a DateField or DateTimeField.",
+            "admin.E128",
+        )
+
 
 class OrderingCheckTests(CheckTestCase):
```
</details>

john-parton replied:
I believe this should work even with a 'virtual' column

Should I add tests for `supports_virtual_generated_columns` as well?

john-parton replied:
@sarahboyce I took the example test cases you provided and fleshed them out so that it will run for both the db features `supports_stored_generated_columns` and `supports_virtual_generated_columns`, setting `db_persist` to the appropriate value.

In the future, if there's a database backend that supports BOTH features, then the test will only be run for the db_persist == False case.

If we would rather, I can create separate methods for each case and have them be thin wrappers around another assert statement.

Maybe something like this

```python
def assertGeneratedFieldIsValid(self, *, db_persist):

    class TestModel(Model):
        date = models.DateTimeField()
        date_copy = models.GeneratedField(
            expression=F("date"),
            output_field=models.DateTimeField(),
            db_persist=db_persist,
        )

    class TestModelAdmin(ModelAdmin):
        date_hierarchy = "date_copy"

    self.assertIsValid(TestModelAdmin, TestModel)

@skipUnlessDBFeature("supports_stored_generated_columns")
@isolate_apps("modeladmin")
def test_valid_case_virtual_generated_field(self):
    self.assertGeneratedFieldIsValid(db_persist=True)

@skipUnlessDBFeature("supports_virtual_generated_columns")
@isolate_apps("modeladmin")
def test_valid_case_stored_generated_field(self):
    self.assertGeneratedFieldIsValid(db_persist=False)
```

Let me know if that general approach is preferred and I'll update the pull request.

sarahboyce replied:
I like the `assertGeneratedFieldIsValid` pattern :heart_eyes: 
This has been something I've found quite difficult to manage in tests in general

We might be ok just having the one though as it is enough to give us coverage for the `GeneratedField` case (Simon is more familiar with ORM test handling than me so he might have an opinion also)

john-parton replied:
I think this version is actually more readable and technically more "correct". Having the random "assert" in the previous version was a bad smell.

john-parton commented:
I added a release note.

Is there a way that I can render out those pages locally so that I can make sure they're formatted nicely?

I don't think it's necessary to add a 'version changed' note to the date_hierarchy docs. I think a typical user of the `GeneratedField` will just assume it's going to work.

sarahboyce commented:
> I added a release note.

Thank you! This is a special case where the release note belongs in the 5.0.8 release notes (as this is a release blocker for a monthly release of 5.0). Something like
```
* Fixed a bug in Django 5.0 where a system check would crash when
  ``ModelAdmin.date_hierarchy`` was set to a ``GeneratedField`` with an
  ``output_field`` of a ``DateField`` or a ``DateTimeField``
  (:ticket:`35628`).
```
You might need to rebase with the latest main

> Is there a way that I can render out those pages locally so that I can make sure they're formatted nicely?

Yes, see [this guide](https://docs.djangoproject.com/en/5.0/internals/contributing/writing-documentation/#build-the-documentation-locally) also have the requirements of `docs/requirements.txt` installed :+1: 

> I don't think it's necessary to add a 'version changed' note to the date_hierarchy docs. I think a typical user of the `GeneratedField` will just assume it's going to work.

Correct :+1:

john-parton commented:
- Rewrote the tests using the assertion pattern discussed previously
- Moved release note to 5.0.8 bugfixes section
- Squashed my commit and cherry-picked it against main

john-parton commented:
Fixed linter errors, squashed, force-pushed.

sarahboyce reviewed (approved):
Thank you @john-parton :star: this looks good to me

pauloxnet commented:
LGTM

charettes commented on tests/modeladmin/test_checks.py:
```diff
@@ -97,6 +98,30 @@ class TestModelAdmin(ModelAdmin):
 
         self.assertIsValid(TestModelAdmin, ValidationTestModel)
 
+    def assertGeneratedDateTimeFieldIsValid(self, *, db_persist):
```
Any thoughts about decorating `assertGeneratedDateTimeFieldIsValid` with `isolate_apps` instead of the callers since it's the one that actually creates and registers a model?

```suggestion
    @isolate_apps("modeladmin")
    def assertGeneratedDateTimeFieldIsValid(self, *, db_persist):
```

sarahboyce replied:
Good idea :+1:

charettes commented on tests/modeladmin/test_checks.py:
```diff
@@ -97,6 +98,30 @@ class TestModelAdmin(ModelAdmin):
+        self.assertIsValid(TestModelAdmin, TestModel)
+
+    @skipUnlessDBFeature("supports_stored_generated_columns")
+    @isolate_apps("modeladmin")
```
```suggestion
```

charettes commented on tests/modeladmin/test_checks.py:
```diff
@@ -97,6 +98,30 @@ class TestModelAdmin(ModelAdmin):
+        self.assertGeneratedDateTimeFieldIsValid(db_persist=True)
+
+    @skipUnlessDBFeature("supports_virtual_generated_columns")
+    @isolate_apps("modeladmin")
```
```suggestion
```

charettes commented on tests/modeladmin/test_checks.py:
```diff
@@ -1029,6 +1054,34 @@ class TestModelAdmin(ModelAdmin):
+        )
+
+    @skipUnlessDBFeature("supports_stored_generated_columns")
+    @isolate_apps("modeladmin")
```
```suggestion
```

charettes commented on tests/modeladmin/test_checks.py:
```diff
@@ -1029,6 +1054,34 @@ class TestModelAdmin(ModelAdmin):
+        self.assertGeneratedIntegerFieldIsInvalid(db_persist=True)
+
+    @skipUnlessDBFeature("supports_virtual_generated_columns")
+    @isolate_apps("modeladmin")
```
```suggestion
```

charettes commented on tests/modeladmin/test_checks.py:
```diff
@@ -1029,6 +1054,34 @@ class TestModelAdmin(ModelAdmin):
             "admin.E128",
         )
 
+    def assertGeneratedIntegerFieldIsInvalid(self, *, db_persist):
```
```suggestion
    @isolate_apps("modeladmin")
    def assertGeneratedIntegerFieldIsInvalid(self, *, db_persist):
```
